drivers/led/lp503x.c

```diff
@@ -0,0 +1,286 @@
+#include <device.h>
+#include <zephyr.h>
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(lp503x);
+
+#define LP503X_DEVICE_CONFIG0		0
+#define   CONFIG0_CHIP_EN		(1 << 6)
+
+#define LP503X_DEVICE_CONFIG1		0x1
+#define   CONFIG1_LED_GLOBAL_OFF	(1 << 0)
+#define   CONFIG1_MAX_CURRENT_OPT	(1 << 1)
+#define   CONFIG1_PWM_DITHERING_EN	(1 << 2)
+#define   CONFIG1_AUTO_INCR_EN		(1 << 3)
+#define   CONFIG1_POWER_SAVE_EN		(1 << 4)
+#define   CONFIG1_LOG_SCALE_EN		(1 << 5)
+
+#define LP503X_LED_CONFIG0		0x2
+#define   CONFIG0_LED0_BANK_EN		(1 << 0)
+#define   CONFIG0_LED1_BANK_EN		(1 << 1)
+#define   CONFIG0_LED2_BANK_EN		(1 << 2)
+#define   CONFIG0_LED3_BANK_EN		(1 << 3)
+#define   CONFIG0_LED4_BANK_EN		(1 << 4)
+#define   CONFIG0_LED5_BANK_EN		(1 << 5)
+#define   CONFIG0_LED6_BANK_EN		(1 << 6)
+#define   CONFIG0_LED7_BANK_EN		(1 << 7)
+
+#define LP503X_LED_CONFIG1		0x3
+#define   CONFIG1_LED8_BANK_EN		(1 << 0)
+#define   CONFIG1_LED9_BANK_EN		(1 << 1)
+#define   CONFIG1_LED10_BANK_EN		(1 << 2)
+#define   CONFIG1_LED11_BANK_EN		(1 << 3)
+
+#define LP503X_BANK_BRIGHTNESS		0x4
+#define LP503X_BANK_A_COLOR		0x5
+#define LP503X_BANK_B_COLOR		0x6
+#define LP503X_BANK_C_COLOR		0x7
+
+#define LP503X_LED_BRIGHTNESS_BASE	0x8
+#define LP503X_OUT_COLOR_BASE		0x14
+
+/* Expose channels starting from the bank registers. */
+#define LP503X_CHANNEL_BASE		LP503X_BANK_BRIGHTNESS
+
+#define DEV_CFG(dev)  ((const struct lp503x_config *) \
+		       ((dev)->config_info))
+#define DEV_DATA(dev) ((struct lp503x_data *) \
+		       ((dev)->driver_data))
+
+struct lp503x_config {
+	char *i2c_bus_label;
+	uint8_t i2c_addr;
+	uint8_t num_leds;
+	bool log_scale_en;
+	bool max_curr_opt;
+	const struct led_info *leds_info;
+};
+
+struct lp503x_data {
+	struct device *i2c;
+	uint8_t *chan_buf;
+};
+
+static const struct led_info *
+lp503x_led_to_info(const struct lp503x_config *config, uint32_t led)
+{
+	int i;
+
+	for (i = 0; i < config->num_leds; i++) {
+		if (config->leds_info[i].index == led) {
+			return &config->leds_info[i];
+		}
+	}
+	return NULL;
+}
+
+static int lp503x_get_info(struct device *dev, uint32_t led,
+			   const struct led_info **info)
+{
+	const struct lp503x_config *config = DEV_CFG(dev);
+	const struct led_info *led_info = lp503x_led_to_info(config, led);
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	*info = led_info;
+
+	return 0;
+}
+
+static int lp503x_set_brightness(struct device *dev,
+				 uint32_t led, uint8_t value)
+{
+	const struct lp503x_config *config = DEV_CFG(dev);
+	struct lp503x_data *data = DEV_DATA(dev);
+	const struct led_info *led_info = lp503x_led_to_info(config, led);
+	uint8_t buf[2];
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	buf[0] = LP503X_LED_BRIGHTNESS_BASE + led_info->index;
+	buf[1] = value;
+
+	return i2c_write(data->i2c, buf, sizeof(buf), config->i2c_addr);
+}
+
+static int lp503x_on(struct device *dev, uint32_t led)
+{
+	return lp503x_set_brightness(dev, led, 255);
```

Comment:
Brightness is documented as being in percent in led.h. Which one is right?

Reply:
Yes you are right. I assumed wrongly that the maximum brightness value was 255, as in Linux. That being said maybe it would make sense to allow led_set_brightness to take a brightness parameter going up to max_brightness (with max_brightness stored in struct led_info). Looking at the other LED drivers it seems they are all already handling the brightness parameter as a value being in the [min_brightness-max_brightness] range, and not as a percentage. Note that they all conveniently set max_brightness to 100, even when the hardware value is different...

Reply:
I think that having an API that doesn't sacrifice precision when setting brightness makes perfect sense. However, what you are proposing is an incompatible stable API change (since led.h is declared stable, and this would require application source code which relied on this being in percent to change).

Changing a stable API incompatibly is a lot more work, as documented here:

https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#introducing-incompatible-changes

If you are up for jumping through those hoops, I think it makes sense, but we'd have to see what the entire community thinks. If you want to move that change out of the scope of this initial PR, to be addressed separately, in my opinion that might be a better way forward, but it's your call.

Reply:
Yes it is definitely out of the scope of this PR. I'll come back at this.